src/main.cpp

```diff
@@ -2785,6 +2848,113 @@ bool AbortNode(const std::string &strMessage, const std::string &userMessage) {
     return false;
 }
 
+
+/**
+ * BLOCK PRUNING CODE
+ */
+
+/* Calculate the amount of disk space the block & undo files currently use */
+uint64_t CalculateCurrentUsage()
+{
+    uint64_t retval = 0;
+    BOOST_FOREACH(const CBlockFileInfo &file, vinfoBlockFile) {
+        retval += file.nSize + file.nUndoSize;
+    }
+    return retval;
+}
+
+/* Prune a block file (modify associated database entries)*/
+void PruneOneBlockFile(const int fileNumber)
+{
+    for (BlockMap::iterator it = mapBlockIndex.begin(); it != mapBlockIndex.end(); ++it) {
```

Comment:
I wrote an in-memory reverse lookup, which gets populated on startup.  On my computer the brute force way takes about 20ms, whereas the reverse lookup takes 200us.  The problem is the size of the in-memory structure, it will only get populated if you're in prune mode, but if you're keeping the whole current blockchain for instance, I think its about 10MB.   I decided to punt on the problem for now.